vhost_user_backend: Add new crate

sboeuf opened the pull request:
The purpose of this new crate is to provide a common library to all
vhost-user backend implementations. The more is handled by this library,
the less duplication will need to happen in each vhost-user daemon.

This crate relies a lot on vhost_rs, vm-memory and vm-virtio crates.

Signed-off-by: Sebastien Boeuf <sebastien.boeuf@intel.com>

sboeuf commented:
@sameo @rbradford @bjzhjing @yangzhon 
Here is the first draft of the libvhost-user in Rust. It still needs to be improved with proper error checking and maybe some improvements reducing the need for mutex, but I wanted to get a first review on it before moving further.

@bjzhjing @yangzhon could you please try to adapt your code (`net` and `blk` respectively) so that it would rely on this library? Also, I'm sure there's a ton of improvement needed, so please point to me anything that needs to change in order to support real backend implementations.

rbradford commented:
This looks like a good start to me. I think tarting it off inside the cloud-hypervisor repo is a good idea whilst it matures and the backends that use it are developed.

sboeuf commented:
@rbradford thanks for reviewing!
I have updated the code by cleaning a few things. I added commits to make clear what the changes are. Let me know what you think.

rbradford commented on Cargo.toml:
```diff
@@ -27,3 +27,6 @@ mmio = ["vmm/mmio_support"]
 
 [workspace]
+members = [
+    "vhost_user_backend",
```
Not sure about this? I think we should have them all or nothing.

sboeuf replied:
Yes that was the wrong way to do it. I've just fixed by making the crate its own workspace.
